mpOpenAPI module selection further tests

Azquelt opened the pull request:
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Fixes #29874

Azquelt committed:
mpOpenAPI: Adapt existing merge config tests

Adapt the existing merge config tests which we have for MP Config so
that they test the same scenarios with server.xml config.

Azquelt committed:
mpOpenAPI: refactor include config update test

Move and streamline the tests which test updating the module selection
config while apps are deployed.

Being in the same test class as the other config update tests will mean
one fewer server to start and the ability to reuse the supporting
framework code.

Azquelt committed:
mpOpenAPI: Fix testMPConfigIgnoredInServerXMLMode

Move this test to its own class because it needs to set MP Config
properties before the server starts to work properly.

Use correct config elements in OpenAPIMergeTestServer server.xml.

Azquelt committed:
mpOpenAPI: refactor MergeServerXMLTest away

Move the tests from MergeServerXMLTest into MergeConfigServerXMLTest.

MergeConfigServerXMLTest now includes equivalents to all the tests from
MergeConfigTest, plus the additional tests specific to server.xml config
previously in MergeServerXMLTest.

Having all the tests in one class makes it easier to share setup code
and reduces the number of server restarts required.

Azquelt commented:
#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.openapi.1.1_fat_tck,com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.microprofile.openapi_fat_tck,com.ibm.ws.openapi.ui_fat,io.openliberty.microprofile.openapi.2.0.internal_fat,io.openliberty.microprofile.openapi.2.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.1.internal_fat_tck,io.openliberty.microprofile.openapi.4.0.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat_tck,io.openliberty.microprofile.openapi.ui.internal_fat

benjamin-confino commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/utils/OpenAPITestUtil.java:
```diff
@@ -277,6 +279,19 @@ public static void checkPaths(JsonNode root,
         assertThat("Path names", pathNames, hasSize(expectedCount));
     }
 
+    public static void checkServerContextRoots(JsonNode root, String expectedContextRoot) {
```
This method is called `...ContextRoots` but only takes a singular context root. Could be fixed in either direction.

Azquelt replied:
It checks every server's path has the expected context root, so multiple servers are checked for the same context root.

I guess `checkServersForContextRoot` would be clearer.

Azquelt replied:
Changed the name and added some doc.

benjamin-confino commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/MergeConfigServerXMLTest.java:
```diff
@@ -63,18 +68,295 @@ public class MergeConfigServerXMLTest {
     @AfterClass
     public static void shutdown() throws Exception {
-        server.stopServer("CWWKO1683W"); // Invalid info element
+        server.stopServer("CWWKO1683W" // Invalid info element
```
The comment is inside the method parentheses.

Azquelt replied:
Yeah, this was deliberate because I meant to add more messages here and each message wants its own comment.

benjamin-confino replied:
And now I see more messages, with more of them it makes sense.

benjamin-confino commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/MergeConfigServerXMLTest.java:
```diff
@@ -63,18 +68,295 @@ public class MergeConfigServerXMLTest {
+    }
+
+    @Test
+    public void testDefaultInclusion() throws Exception {
```
A lot of tests are duplicated between here and MergeConfigTest with the only difference being weather the config is set via mpConfig or serverXML. Consider opening a tech debt item to make both extend a common abstract class?

Azquelt replied:
We could do, I think it's borderline. There's a good bit that's the same, but there's a fair bit that's different in the setup, in exactly what we check for and in which mode each test runs in.

I think on balance it's going to be easier to follow what the test is doing if everything's in the same class, rather than having to bounce between subclasses to work out what's going on.

benjamin-confino commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/MergeConfigServerXMLTest.java:
```diff
@@ -63,18 +68,295 @@ public class MergeConfigServerXMLTest {
+        OpenAPITestUtil.checkPaths(openapiNode, 1, "/test2");
+
+        // Test that merging disabled message was output (at some point)
+        assertThat(server.findStringsInLogs(" I CWWKO1663I:.*Combining OpenAPI documentation from multiple modules is disabled."),
```
The equivalent line in MergeConfigTest ends in `.*test1` should it here?

Azquelt replied:
The error message here should include the application name. However, the error message is only output once for each server start (this message appears when you deploy an additional app and we don't want to spam it if the user deploys lots of apps).

In `MergeConfigTest` the server is restarted for each test so each test will get its own instance of the message. In this test the server is started once for the whole class, so the message is only output when the first test that prompts it runs.

Since we don't know the test order, we don't know what application name will be in the message. When the second test runs, it's finding the message that was output when the first test ran.

benjamin-confino commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/MergeConfigServerXMLTest.java:
```diff
@@ -63,18 +68,295 @@ public class MergeConfigServerXMLTest {
+        OpenAPITestUtil.checkPaths(openapiNode, 1);
+
+        // Test that merging disabled message was output (at some point)
+        assertThat(server.findStringsInLogs(" I CWWKO1663I:.*Combining OpenAPI documentation from multiple modules is disabled."),
```
The equivalent line in MergeConfigTest ends in `.*test1` should it here?

Azquelt committed:
Fix a race condition in mpOpenAPI tests

These tests could fail because we were restoring the server.xml config
and then deleting apps from dropins.

Restoring the config could cause mpOpenAPI to start scanning
applications which could fail if the app archive was deleted when this
was happenning.

benjamin-confino commented on dev/io.openliberty.microprofile.openapi.2.0.internal_fat/fat/src/io/openliberty/microprofile/openapi20/fat/deployments/MergeConfigServerXMLTest.java:
```diff
@@ -63,18 +68,300 @@ public class MergeConfigServerXMLTest {
+            OpenAPITestUtil.checkPaths(openapiNode, 1, "/test");
+
+            // Test that a message was output
+            assertThat(server.findStringsInLogs(" I CWWKO1663I:.*Combining OpenAPI documentation from multiple modules is disabled.*test1"),
```
As per a previous discussion about CWWKO1663I. We should remove test1 from here

Azquelt replied:
Done

Azquelt commented:
#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.openapi.1.1_fat_tck,com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.microprofile.openapi_fat_tck,com.ibm.ws.openapi.ui_fat,io.openliberty.microprofile.openapi.2.0.internal_fat,io.openliberty.microprofile.openapi.2.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.1.internal_fat_tck,io.openliberty.microprofile.openapi.4.0.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat_tck,io.openliberty.microprofile.openapi.ui.internal_fat
